src/python/pants/backend/scala/bsp/rules.py

```diff
@@ -0,0 +1,120 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+from dataclasses import dataclass
+
+from pants.backend.scala.bsp.spec import (
+    ScalaBuildTarget,
+    ScalacOptionsItem,
+    ScalacOptionsParams,
+    ScalacOptionsResult,
+    ScalaPlatform,
+)
+from pants.backend.scala.dependency_inference.symbol_mapper import AllScalaTargets
+from pants.backend.scala.subsystems.scala import ScalaSubsystem
+from pants.base.build_root import BuildRoot
+from pants.bsp.rules import BSPBuildTargets, BSPBuildTargetsRequest
+from pants.bsp.spec import BuildTarget, BuildTargetCapabilities, BuildTargetIdentifier
+from pants.build_graph.address import AddressInput
+from pants.engine.internals.selectors import Get, MultiGet
+from pants.engine.rules import QueryRule, collect_rules, rule
+from pants.engine.target import Target, WrappedTarget
+from pants.engine.unions import UnionRule
+from pants.jvm.subsystems import JvmSubsystem
+from pants.jvm.target_types import JvmResolveField
+
+
+class ScalaBSPBuildTargetsRequest(BSPBuildTargetsRequest):
+    pass
+
+
+def _pants_target_to_bsp_build_target(
+    target: Target, jvm: JvmSubsystem, scala: ScalaSubsystem
+) -> BuildTarget:
+    resolve = target[JvmResolveField].normalized_value(jvm)
```

Comment:
This is imo a tricky assumption -- you must pass in a certain type of target but that isn't made clear with the signature. It suggests any `Target` works.

Instead, you could have the user pass `JvmResolveField` directly. Note that it has `AsyncFieldMixin`, so you can access its `.address` still. That will make it clear at callers what targets are expected.